optional_plugins/loader_results_importer/avocado_loader_results_importer/__init__.py

```diff
@@ -0,0 +1,231 @@
+
+        tests = []
+        with open(url, 'r') as results:
+            no_tests = None
+            _test = None
+            _content = []
+            try:
+                for line in results:
+                    if no_tests is None:
+                        no_tests = self.__re_no_tests.match(line)
+                        if no_tests:
+                            no_tests = no_tests.group(1)
+                            continue
+                    match = self.__re_test_result.match(line)
+                    if match:
+                        if _test:
+                            self.add_test(tests, _test, _content)
+                        _content = []
+                        _test = match.groups()
+                    else:
+                        _content.append(line[:-1])
+            except ValueError as details:
+                if which_tests is loader.DEFAULT:
+                    return []
+                return [(ResultsImporterTest, {"name": "%s: %s"
+                                                       % (reference,
+                                                          details)})]
+        if _test:
+            self.add_test(tests, _test, _content)
+        if no_tests is None:
+            if which_tests is loader.DEFAULT:
+                return []
+            return [(ResultsImporterTest, {"name": "%s: Number of tests not "
+                                                   "specified in TAP file."
+                                                   % reference})]
+        for i, test in enumerate(tests):
+            if test is None:
+                kwargs = {"name": "Missing",
+                          "results_importer_state": {"status": ERROR}}
+                tests[i] = (ResultsImporterTest, kwargs)
+        return tests
+
+
+class ResultsImporterLoader(loader.TestLoader):
+
+    """
+    Uses $type://$url format to find results in supported $type and loads
+    them as tests so they can be integrated into overall results.
+    """
+
+    name = "results_importer"
+    _tap_loader = None
+
+    @staticmethod
+    def get_type_label_mapping():
+        """
+        No type is discovered by default, uses "full_*_mappings" to report
+        the actual types after "discover()" is called.
+        """
+        return {ResultsImporterTest: 'IMPORTED',
+                NotResultsImporterTest: '!IMPORTED'}
+
+    @staticmethod
+    def get_decorator_mapping():
+        return {ResultsImporterTest: output.TERM_SUPPORT.healthy_str,
+                NotResultsImporterTest: output.TERM_SUPPORT.fail_header_str}
+
+    def discover(self, reference, which_tests=loader.DEFAULT):
+        if reference is None:
+            return []
+        if reference.startswith("tap://"):
+            if self._tap_loader is None:
+                self._tap_loader = TapDiscoverer()
+            return self._tap_loader.discover(reference, which_tests)
+        return []
```

Comment:
Well, you're parsing a TAP output into an Avocado job results without actually executing anything. Avocado is acting as a parse tool. I'm not sure there's value in that, because in my understanding, the use case is to execute a (SIMPLE) test which outputs the results TAP format and, based on that TAP output, populate the results, all in the same Job.

Example::
```
$ cat /tmp/test.sh 
#!/bin/bash
/home/apahim/git/qemu/tests/ahci-test --tap
```
Current behavior:
```
$ avocado run /tmp/test.sh 
JOB ID     : 6c472806d0b65776606766c4ad4a4c66f73d1396
JOB LOG    : /home/apahim/avocado/job-results/job-2018-02-15T10.22-6c47280/job.log
 (1/1) /tmp/test.sh: PASS (5.46 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 5.56 s
JOB HTML   : /home/apahim/avocado/job-results/job-2018-02-15T10.22-6c47280/results.html
```

Expected behavior:

```
$ avocado run --load-results tap -- /tmp/test.sh
JOB ID     : a332b251ad450e318f948be7c938ae68f20b32ad
JOB LOG    : /home/apahim/avocado/job-results/job-2018-02-15T10.13-a332b25/job.log
 (01/66)  /x86_64/ahci/sanity: PASS (0.04 s)
 (02/66)  /x86_64/ahci/pci_spec: PASS (0.04 s)
 (03/66)  /x86_64/ahci/pci_enable: PASS (0.05 s)
 (04/66)  /x86_64/ahci/hba_spec: PASS (0.04 s)
 (05/66)  /x86_64/ahci/hba_enable: PASS (0.05 s)
 (06/66)  /x86_64/ahci/identify: PASS (0.03 s)
 (07/66)  /x86_64/ahci/max: PASS (0.03 s)
 (08/66)  /x86_64/ahci/reset: PASS (0.02 s)
 (09/66)  /x86_64/ahci/io/pio/lba28/simple/zero: PASS (0.02 s)
 (10/66)  /x86_64/ahci/io/pio/lba28/simple/low: PASS (0.02 s)
 (11/66)  /x86_64/ahci/io/pio/lba28/simple/high: PASS (0.02 s)
 (12/66)  /x86_64/ahci/io/pio/lba28/double/zero: PASS (0.02 s)
 (13/66)  /x86_64/ahci/io/pio/lba28/double/low: PASS (0.02 s)
 (14/66)  /x86_64/ahci/io/pio/lba28/double/high: PASS (0.02 s)
 (15/66)  /x86_64/ahci/io/pio/lba28/long/zero: PASS (0.03 s)
 (16/66)  /x86_64/ahci/io/pio/lba28/long/low: PASS (0.02 s)
 (17/66)  /x86_64/ahci/io/pio/lba28/long/high: PASS (0.02 s)
 (18/66)  /x86_64/ahci/io/pio/lba28/short/zero: PASS (0.03 s)
 (19/66)  /x86_64/ahci/io/pio/lba28/short/low: PASS (0.02 s)
 (20/66)  /x86_64/ahci/io/pio/lba28/short/high: PASS (0.02 s)
 (21/66)  /x86_64/ahci/io/pio/lba48/simple/zero: PASS (0.02 s)
 (22/66)  /x86_64/ahci/io/pio/lba48/simple/low: PASS (0.02 s)
 (23/66)  /x86_64/ahci/io/pio/lba48/double/zero: PASS (0.02 s)
 (24/66)  /x86_64/ahci/io/pio/lba48/double/low: PASS (0.02 s)
 (25/66)  /x86_64/ahci/io/pio/lba48/long/zero: PASS (0.02 s)
 (26/66)  /x86_64/ahci/io/pio/lba48/long/low: PASS (0.03 s)
 (27/66)  /x86_64/ahci/io/pio/lba48/short/zero: PASS (0.02 s)
 (28/66)  /x86_64/ahci/io/pio/lba48/short/low: PASS (0.02 s)
 (29/66)  /x86_64/ahci/io/dma/lba28/fragmented: PASS (0.03 s)
 (30/66)  /x86_64/ahci/io/dma/lba28/retry: PASS (0.03 s)
 (31/66)  /x86_64/ahci/io/dma/lba28/simple/zero: PASS (0.03 s)
 (32/66)  /x86_64/ahci/io/dma/lba28/simple/low: PASS (0.03 s)
 (33/66)  /x86_64/ahci/io/dma/lba28/simple/high: PASS (0.03 s)
 (34/66)  /x86_64/ahci/io/dma/lba28/double/zero: PASS (0.03 s)
 (35/66)  /x86_64/ahci/io/dma/lba28/double/low: PASS (0.03 s)
 (36/66)  /x86_64/ahci/io/dma/lba28/double/high: PASS (0.02 s)
 (37/66)  /x86_64/ahci/io/dma/lba28/long/zero: PASS (0.02 s)
 (38/66)  /x86_64/ahci/io/dma/lba28/long/low: PASS (0.03 s)
 (39/66)  /x86_64/ahci/io/dma/lba28/long/high: PASS (0.03 s)
 (40/66)  /x86_64/ahci/io/dma/lba28/short/zero: PASS (0.04 s)
 (41/66)  /x86_64/ahci/io/dma/lba28/short/low: PASS (0.03 s)
 (42/66)  /x86_64/ahci/io/dma/lba28/short/high: PASS (0.03 s)
 (43/66)  /x86_64/ahci/io/dma/lba48/simple/zero: PASS (0.02 s)
 (44/66)  /x86_64/ahci/io/dma/lba48/simple/low: PASS (0.02 s)
 (45/66)  /x86_64/ahci/io/dma/lba48/double/zero: PASS (0.02 s)
 (46/66)  /x86_64/ahci/io/dma/lba48/double/low: PASS (0.03 s)
 (47/66)  /x86_64/ahci/io/dma/lba48/long/zero: PASS (0.03 s)
 (48/66)  /x86_64/ahci/io/dma/lba48/long/low: PASS (0.03 s)
 (49/66)  /x86_64/ahci/io/dma/lba48/short/zero: PASS (0.03 s)
 (50/66)  /x86_64/ahci/io/dma/lba48/short/low: PASS (0.05 s)
 (51/66)  /x86_64/ahci/io/ncq/simple: PASS (0.04 s)
 (52/66)  /x86_64/ahci/io/ncq/retry: PASS (0.05 s)
 (53/66)  /x86_64/ahci/flush/simple: PASS (0.03 s)
 (54/66)  /x86_64/ahci/flush/retry: PASS (0.02 s)
 (55/66)  /x86_64/ahci/flush/migrate: PASS (0.02 s)
 (56/66)  /x86_64/ahci/migrate/sanity: PASS (0.03 s)
 (57/66)  /x86_64/ahci/migrate/dma/simple: PASS (0.02 s)
 (58/66)  /x86_64/ahci/migrate/dma/halted: PASS (0.02 s)
 (59/66)  /x86_64/ahci/migrate/ncq/simple: PASS (0.02 s)
 (60/66)  /x86_64/ahci/migrate/ncq/halted: PASS (0.02 s)
 (61/66)  /x86_64/ahci/cdrom/eject: PASS (0.03 s)
 (62/66)  /x86_64/ahci/cdrom/dma/single: PASS (0.02 s)
 (63/66)  /x86_64/ahci/cdrom/dma/multi: PASS (0.03 s)
 (64/66)  /x86_64/ahci/cdrom/pio/single: PASS (0.02 s)
 (65/66)  /x86_64/ahci/cdrom/pio/multi: PASS (0.03 s)
 (66/66)  /x86_64/ahci/cdrom/pio/bcl: PASS (0.02 s)
RESULTS    : PASS 66 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 5.62 s
JOB HTML   : /home/apahim/avocado/job-results/job-2018-02-15T10.13-a332b25/results.html
```

Reply:
Yes, I know Paolo was mainly interested in this, but we discussed 2 different use-cases, the later is not really possible without concept changes while the first one (this one) is quite simple and might serve well for some use-cases.

For example I am executing libguestfs tests which produces (different) results and it'd be nice to easily import them into job results. On the other hand when we have the later, this one would become almost useless (as one can simply run `cp` to get the previously created results). So it all depends whether and in what future do we want to work on the first way.

Btw about the first case I can imagine multiple solutions, some achievable even nowadays (with minor changes to the runner. Sorted according to my preferences:

1. Allowing subtest results in a job - your example would finish with:

    (1/1) /tmp/tests.sh: PASS (5.62)
        * 1/67 /x86_64/ahci/sanity: PASS (0.04 s)
        ...

2. Allowing the test to inject tests into the suite:

    (1/1) /tmp/tests.sh: PASS (5.62)
    (02/67) /x86_64/ahci/sanity: PASS (0.04 s)
    ....

3. The same as (2) but allowing the loader to say dynamic number of tests will be fenerated:

    (0001/x) /tmp/tests.sh: PASS (5.62)
    (0002/x) /x86_64/ahci/sanity: PASS (0.04 s)
    ...

4. Execute the test on load (which I dislike) therefor knowing the no tests in advance:

    (01/67) /x86_64/ahci/sanity: PASS (0.04 s)
    ...

The (1) would require major change in the concept, (2, 3, 4) should be doable eg. via queue or status, but might change some assumptions further on.

Reply:
PS: about the usage I don't think it should be a `--load-results tap` option, but it should be a part of `SimpleTest` to allow to point to the `tap` results or `status`/`queue` command for instrumented tests.

Also the position of the `/tmp/tests.sh` results would be either top, in case everything is executed first, or last, in case we allow subtests to be reported during execution (which would be probably even better and would align to Avocado-vt's `context`.

Reply:
... in such case I'd expect something like:

    (1/1) /tmp/tests.sh: \
      * 1/67 /x86_64/ahci/sanity: PASS (0.04 s)
      ...
    (1/1) /tmp/tests.sh: PASS (5.62)

Reply:
One thing to *not* forget here is **not to try to make the components fit with extra force**.  What I mean is, attempting to only import results, should be seen and implemented just as that.  It makes no sense to have something like:

```
$ avocado run --load-results=tap -- /bin/test.sh
```

**If** `test.sh` is not going to be executed.

Also, what does this mean to `avocado list`? By definition, `list` is supposed to find tests that Avocado can *execute*, not whose execution results can be imported.  Implementing this as a loader seems to be how we can do it right now, with minor infrastructure changes, **but** we'd be introducing the wrong semantics.  We need to plug this into the right place.  There may be the need to create additional interfaces, and in the end, this would mean a different CLI command.  For instance,

```
$ avocado import-result --format=tap results.tap
```

Needs no explanation, and that is a good thing.